rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,294 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 35 - Simplified Node Joining for AWS
+
+## What
+
+EC2 instances will be able to join a teleport cluster without explicitly
+provisioning a join token on the auth server.
+
+In place of a join token, nodes will present one or more signed AWS API
+requests to the auth server. The auth server will send these requests to the
+public AWS API to confirm that:
+1. Because the AWS API accepted the signed request, the node holds valid AWS
+   credentials with permissions for the relevant API endpoints.
+2. Based on the API response, the node's AWS credentials belong to the
+   configured AWS Organization and/or Account.
+
+### tl;dr
+
+Configure auth server to allow nodes to join with the new aws join method:
+```yaml
+teleport:
+  nodename: auth
+
+auth_service:
+  enabled: yes
+  aws_join:
+    allow:
+    - organization: "o-1111111111"
+```
+
+Create ec2 instances in organization "o-1111111111" with a role with the
+following attached policy:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "organizations:DescribeOrganization"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+Configure teleport nodes on ec2 instances:
+```yaml
+teleport:
+  auth_servers:
+    - auth
+  aws_token:
+    credential_source:
+      type: ec2
+    claims:
+    - organization
+
+ssh_service:
+  enabled: yes
+```
+
+Nodes will be authenticated with the new aws join method and be able to join
+the cluster.
+
+## Why
+
+This will make provisioning nodes on ec2 simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+Much inspiration is taken from Vault's [IAM auth method](https://www.vaultproject.io/docs/auth/aws#iam-auth-method).
+
+### Authentication
+
+In place of a join token, nodes will present one or more signed AWS API
+requests to the auth server based on the configured claims:
+- `sts:GetCallerIdentity` for the "identity" claim and
+- `organizations:DescribeOrganization` for the "organization" claim
+
+These are signed HTTP requests which will be serialized and sent to the auth
```

Comment:
How are you going to serialise it? Will just it be a well-formed HTTP request in a byte stream? Or will there be a higher level structure involved?

Reply:
Could do either honestly, the `net/http` package lets us write/read a `http.Request` to/from a HTTP byte stream pretty easily so I'll probably just do that. The bytes would then be a field of a gRPC message holding all of the separate requests